pkg/v1/internal/gzip/zip.go

```diff
@@ -82,15 +83,18 @@ func UnzipReadCloser(r io.ReadCloser) (io.ReadCloser, error) {
 	}, nil
 }
 
-// Is detects whether the input stream is compressed.
-func Is(r io.Reader) (bool, error) {
-	magicHeader := make([]byte, 2)
-	n, err := r.Read(magicHeader)
-	if n == 0 && err == io.EOF {
-		return false, nil
-	}
+// Is detects whether the input stream is gzip compressed. A small number
+// of bytes are buffered to Peek at the gzip header so that the returned
+// io.Reader can be used as a replacement for the consumed input io.Reader.
+func Is(r io.Reader) (bool, *bufio.Reader, error) {
```

Comment:
Thinking about this a little more, it is kind of cumbersome to use, I'm  okay with merging this as-is for a quick win, but what do you think about some kind of interface like this that hides all of it:

```go
type Reopener struct {
  consumed int32
  gzipped bool
  rc io.ReadCloser
  opener Opener
}

func (ro *Reopener) Open() (io.ReadCloser, error) {
  if atomic.CompareAndSwapInt32(&ro.consumed, 0, 1) {
    return ro.rc, nil
  }
  return rc.opener()
}

func (ro  *Reopener) IsGzipped() bool {
  return ro.gzipped
}

func Reopen(opener Opener) (*Reopener, error) {
  rc, err := opener()
  if err != nil {
    return nil, false err
  }

  br, ok, err := Is(rc)
  if err != nil {
    return false, nil, err
  }

  brc := &and.ReadCloser{
    Reader:    r,
    CloseFunc: rc.Close,
  }

  return &Reopener{
    gzipped: ok,
    rc: rc,
    opener: opener,
  }, nil
}
```

And in `tarball` we can just:

```go
reopener, err := ggzip.Reopen(rc)
if err != nil {
  return nil, err
}

compressed := reopener.IsGzipped()
opener = reopener.Open
```

Names are hard, so I hate them, and perhaps the "reopener" stuff doesn't make sense within the ggzip package because "Opener" is a `tarball` abstraction...

Maybe `ggzip` should just expose something that compares a `[]byte` to `gzipMagicHeader` or even just exposes that array as `ggzip.MagicHeader`? I'm not sure, but having `gzip.Is` take an `io.Reader` certainly feels goofier the more I look at this PR.